### DISCUSSION TITLE
Create Custom PyTreeDef objects, or adapt existing ones.

### DISCUSSION THREAD

**Discussion by peterroelants** [jax-ml/jax#13768] [Q&A] [answered]
Is it possible to create custom `PyTreeDef` objects, or adapting the structure of existing ones?

Context:
I'm stacking custom PyTree objects using `jax.tree_map(lambda *xs: jnp.stack(xs, axis=0), *pytrees)`, however, if these objects contain metadata that are not leaf nodes, but are part of the `PyTreeDef` this fails. 

The following example based on [Treeo](https://cgarciae.github.io/treeo/) fails:
```python
from dataclasses import dataclass

import jax
import jax.numpy as jnp
import treeo as to

@dataclass
class Person(to.Tree):
    height: jnp.array = to.field(node=True) # I am a node field!
    age_static: jnp.array = to.field(node=False) # I am a static field!, I should not be updated.
    name: str = to.field(node=False) # I am a static field!

persons = [
    Person(height=jnp.array(1.8), age_static=jnp.array(25.), name="John"),
    Person(height=jnp.array(1.7), age_static=jnp.array(100.), name="Wald"),
    Person(height=jnp.array(2.1), age_static=jnp.array(50.), name="Karen")
]

# Stack (struct of arrays instead of list of structs)
jax.tree_map(lambda *values: jnp.stack(values, axis=0), *persons)
```

<details>
<summary>
This code results in this exception (open dropdown to see). 

See also: https://github.com/cgarciae/treeo/issues/23

</summary>
<br>

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 18
     11     name: str = to.field(node=False) # I am a static field!
     13 persons = [
     14     Person(height=jnp.array(1.8), age_static=jnp.array(25.), name="John"),
     15     Person(height=jnp.array(1.7), age_static=jnp.array(100.), name="Wald"),
     16     Person(height=jnp.array(2.1), age_static=jnp.array(50.), name="Karen")
     17 ]
---> 18 jax.tree_map(lambda *values: jnp.stack(values, axis=0), *persons)

File ~/workspace/lcms_polymer_model/env/env_conda_local/lcms_polymer_model_env/lib/python3.10/site-packages/jax/_src/tree_util.py:199, in tree_map(f, tree, is_leaf, *rest)
    166 """Maps a multi-input function over pytree args to produce a new pytree.
    167 
    168 Args:
   (...)
    196   [[5, 7, 9], [6, 1, 2]]
    197 """
    198 leaves, treedef = tree_flatten(tree, is_leaf)
--> 199 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    200 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

File ~/workspace/lcms_polymer_model/env/env_conda_local/lcms_polymer_model_env/lib/python3.10/site-packages/jax/_src/tree_util.py:199, in <listcomp>(.0)
    166 """Maps a multi-input function over pytree args to produce a new pytree.
    167 
    168 Args:
   (...)
    196   [[5, 7, 9], [6, 1, 2]]
    197 """
    198 leaves, treedef = tree_flatten(tree, is_leaf)
--> 199 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    200 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

ValueError: Mismatch custom node data: {'_field_metadata': {'height': <treeo.types.FieldMetadata object at 0x7fb8b898ba00>, 'age_static': <treeo.types.FieldMetadata object at 0x7fb8b90c0a90>, 'name': <treeo.types.FieldMetadata object at 0x7fb8b8bf9db0>, '_field_metadata': <treeo.types.FieldMetadata object at 0x7fb8b89b56f0>, '_factory_fields': <treeo.types.FieldMetadata object at 0x7fb8b89b5750>, '_default_field_values': <treeo.types.FieldMetadata object at 0x7fb8b89b5660>, '_subtrees': <treeo.types.FieldMetadata object at 0x7fb8b89b5720>}, 'age_static': DeviceArray(25., dtype=float32, weak_type=True), 'name': 'John'} != {'_field_metadata': {'height': <treeo.types.FieldMetadata object at 0x7fb8b898ba00>, 'age_static': <treeo.types.FieldMetadata object at 0x7fb8b90c0a90>, 'name': <treeo.types.FieldMetadata object at 0x7fb8b8bf9db0>, '_field_metadata': <treeo.types.FieldMetadata object at 0x7fb8b89b56f0>, '_factory_fields': <treeo.types.FieldMetadata object at 0x7fb8b89b5750>, '_default_field_values': <treeo.types.FieldMetadata object at 0x7fb8b89b5660>, '_subtrees': <treeo.types.FieldMetadata object at 0x7fb8b89b5720>}, 'age_static': DeviceArray(100., dtype=float32, weak_type=True), 'name': 'Wald'}; value: Person(height=DeviceArray(1.7, dtype=float32, weak_type=True), age_static=DeviceArray(100., dtype=float32, weak_type=True), name='Wald').
```

</details>

I think it's clear to me how I'm getting this exception, it's because the `stack` method is not applied to the static fields (`node=False`). These static fields are stored in the PyTreeDef when flattening the tree.

I could do something like the following:
```python
def stack_pytrees(pytrees):
    tree, *rest = pytrees
    leaves, treedefs = zip(*[jax.tree_util.tree_flatten(tree) for tree in pytrees])
    return treedefs[0].unflatten(jnp.stack(xs, axis=0) for xs in zip(*leaves))


stack_pytrees(persons)
# Person(height=DeviceArray([1.8, 1.7, 2.1], dtype=float32, weak_type=True), age_static=DeviceArray(25., dtype=float32, weak_type=True), name='John')
```
But this removes all metadata from all but the first PyTree. Ideally I would write some custom code mapping the PyTreeDef metadata to a list of that metadata or something. However, I cannot seem to figure out how to create a new custom `PyTreeDef` object where this is the case, or adapt the previous object. Can anyone help me trying to solve this problem (or tell me why I should not be doing this, and should solve this some other way)?

**Comment by jakevdp** [comment]
I've not heard of the `treeo` package, but for this to work on the JAX side I believe you'll need to register your class as a pytree. There is information on this at https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees

There has been discussion surrounding automatic registration of dataclasses as pytrees (see #2371) but currently that registration is not automatic, so you need to use these registration mechanisms.

ALternatively, if you believe this is something that is supposed to happen automatically when you import `treeo`, I'd suggest inquiring in the discussions for that package.

**Reply by peterroelants** [reply]
> I've not heard of the treeo package, but for this to work on the JAX side I believe you'll need to register your class as a pytree.

I'm sorry for not providing any context here, [Treeo](https://cgarciae.github.io/treeo/) provides a `treeo.Tree` class which automatically registers the object as a PyTree, similar to [`flax.struct.PyTreeNode`](https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html#flax.struct.PyTreeNode).

## Issue is not related to Treeo

Now, the issue is not related to Treeo or automatic registration. I'm using Treeo just to have a simple example, it should be irrelevant to my issue.  For example, see this code not using Treeo:
```python
from collections import namedtuple

import jax
import jax.numpy as jnp

from jax.tree_util import (
    register_pytree_node_class
)


@register_pytree_node_class
class RegisteredSpecial:

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __repr__(self):
      return "RegisteredSpecial(x={}, y={})".format(self.x, self.y)

    def tree_flatten(self):
      children = (self.x, )
      aux_data = self.y
      return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        return cls(x=children[0], y=aux_data)


pytrees = [
    RegisteredSpecial(x=1., y=2.),
    RegisteredSpecial(x=3., y=6.),
    RegisteredSpecial(x=4., y=8.)
]
# Stack PyTrees
jax.tree_map(lambda *values: jnp.stack(values, axis=0), *pytrees)
```

<details>
<summary> Which results in a (open dropdown to see full exception)

`ValueError: Mismatch custom node data: 2.0 != 6.0; value: RegisteredSpecial(x=3.0, y=6.0).`  exception because it is not able to merge meta-data returned in `aux_data`, and the `aux_data` is actually different between PyTrees . 

</summary>
<br>

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 37
     27         return cls(x=children[0], y=aux_data)
     32 pytrees = [
     33     RegisteredSpecial(x=1., y=2.),
     34     RegisteredSpecial(x=3., y=6.),
     35     RegisteredSpecial(x=4., y=8.)
     36 ]
---> 37 jax.tree_map(lambda *values: jnp.stack(values, axis=0), *pytrees)

File ~/workspace/lcms_polymer_model/env/env_conda_local/lcms_polymer_model_env/lib/python3.10/site-packages/jax/_src/tree_util.py:199, in tree_map(f, tree, is_leaf, *rest)
    166 """Maps a multi-input function over pytree args to produce a new pytree.
    167 
    168 Args:
   (...)
    196   [[5, 7, 9], [6, 1, 2]]
    197 """
    198 leaves, treedef = tree_flatten(tree, is_leaf)
--> 199 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    200 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

File ~/workspace/lcms_polymer_model/env/env_conda_local/lcms_polymer_model_env/lib/python3.10/site-packages/jax/_src/tree_util.py:199, in <listcomp>(.0)
    166 """Maps a multi-input function over pytree args to produce a new pytree.
    167 
    168 Args:
   (...)
    196   [[5, 7, 9], [6, 1, 2]]
    197 """
    198 leaves, treedef = tree_flatten(tree, is_leaf)
--> 199 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    200 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

ValueError: Mismatch custom node data: 2.0 != 6.0; value: RegisteredSpecial(x=3.0, y=6.0).
```

</details>

I can fix the error by not using `jax.tree_map`:
```python
def stack_pytrees(pytrees):
    tree, *rest = pytrees
    leaves, treedefs = zip(*[jax.tree_util.tree_flatten(tree) for tree in pytrees])
    return treedefs[0].unflatten(jnp.stack(xs, axis=0) for xs in zip(*leaves))
    
stack_pytrees(pytrees)
```
However, this results in:
```
RegisteredSpecial(x=[1. 3. 4.], y=2.0)
```
Which looses all but the first `aux_data` "metatdata").

#### Summary:
The is issue is that I want to "stack" multiple PyTrees into one using `jax.tree_map(lambda *values: jnp.stack(values, axis=0), *persons)`, or a custom function using `tree_flatten` and `unflatten`.
Now, this works for any leaf nodes, but any metadata (`aux_data`) that is incorporated in the PyTreeDef gets lost (except for the one from the first element). I want to write a custom "stacking" function for the PyTreeDef metadata (e.g. by aggregating them in a list), however for this to work I need to either update the PyTreeDef or create a new one.

And this creating or editing of the PyTreeDefs is something I cannot seem to figure out.

**Reply by peterroelants** [reply]
@jakevdp Please let me know if this is not clear, and if it's clear if you want me to rewrite the original question.

**Reply by jakevdp** [reply]
I see - it sounds like you're trying to stack data in incompatible pytrees. The fact that each tree has a different value for the `y` metadata means that the trees are different, and therefore incompatible. If all three `y` values were the same, you wouldn't have this issue because the trees would be compatible.

Can you say more about what you expect the output to be? What do you want to happen with the three `y` values specified in the input trees?

**Reply by peterroelants** [reply]
> The fact that each tree has a different value for the y metadata means that the trees are different, and therefore incompatible.

Yes

> Can you say more about what you expect the output to be? What do you want to happen with the three `y` values specified in the input trees?

I would expect:
```python
pytrees = [
    RegisteredSpecial(x=1., y=2.),
    RegisteredSpecial(x=3., y=6.),
    RegisteredSpecial(x=4., y=8.)
]
pytrees_stacked = stack_pytrees(pytrees)  # Custom function, stacks pytrees into single PyTree
print(f"{pytrees_stacked=}")
print(f"{tree_structure(pytrees_stacked)=}")
```
to output something along the lines of:
```
pytrees_stacked=RegisteredSpecial(x=[1. 3. 4.], y=[2.0, 6.0, 8.0])
tree_structure(pytrees_stacked)=PyTreeDef(CustomNode(RegisteredSpecial[[2.0, 6.0, 8.0]], [*]))
```
Basically, I would expect the result to be the same as if I would have created:
```python
RegisteredSpecial(
    x=jnp.array([1., 3., 4.]), 
    y=[2.0, 6.0, 8.0]
)
```
So no information gets lost.

**Reply by jakevdp** [reply]
OK, thanks for the clarification. The only mechanism to do this would be to make `y` a dynamic child like `x`, rather than static metadata associated with each individual tree. The pytree ilnfrastructure has no mechansim for combining static metadata when combining trees.

So, for example you could achieve roughly what you want by defining your tree flattening/unflattening this way:
```python
    def tree_flatten(self):
      children = (self.x, self.y)
      aux_data = None
      return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
      return cls(x=children[0], y=children[1])
```
If you want the `y` values to be static metadata that is combined in the output tree structure, you'll have to write custom functions to manipulate the tree structures in this way.

**Reply by peterroelants** [reply]
> The only mechanism to do this would be to make `y` a dynamic child like `x`, rather than static metadata associated with each individual tree

This would work for JAX supported types, but wouldn't work for non-valid types such as `str` right?

**Reply by peterroelants** [reply]
> The pytree ilnfrastructure has no mechansim for combining static metadata when combining trees.

So there is no way to manually create or edit objects of type `PyTreeDef`?

**Reply by jakevdp** [reply]
I don't know what you mean by "manually create or edit objects of type `PyTreeDef`".

**Reply by peterroelants** [reply]
> I don't know what you mean by "manually create or edit objects of type `PyTreeDef`".

So when flattening I get a tuple of (leaves, treedefs) of which the leaves are JAX objects and treedefs are `PyTreeDef` objects:
```python
pytrees = [
    RegisteredSpecial(x=1., y=2.),
    RegisteredSpecial(x=3., y=6.),
    RegisteredSpecial(x=4., y=8.)
]

leaves, treedefs = zip(*[jax.tree_util.tree_flatten(tree) for tree in pytrees])
print(f"{leaves=}")  # leaves=([1.0], [3.0], [4.0])
print(f"{treedefs=}")  # treedefs=(PyTreeDef(CustomNode(RegisteredSpecial[2.0], [*])), PyTreeDef(CustomNode(RegisteredSpecial[6.0], [*])), PyTreeDef(CustomNode(RegisteredSpecial[8.0], [*])))
print(f"{treedefs[0]}")  # PyTreeDef(CustomNode(RegisteredSpecial[2.0], [*]))
print(f"{type(treedefs[0])}")  # <class 'jaxlib.xla_extension.pytree.PyTreeDef'>
```

It's these `jaxlib.xla_extension.pytree.PyTreeDef` that I want to work with and edit or create.

**Reply by jakevdp** [reply]
Oh, OK. As far as I know, the only supported mechanism for creating PyTree objects is by using `tree_util.tree_flatten`. Perhaps you can use that in the "manual" creation process that you have in mind?

**Reply by peterroelants** [reply]
> Oh, OK. As far as I know, the only supported mechanism for creating PyTree objects is by using `tree_util.tree_flatten`. 

I see. I also found the definition of [`PyTreeDef` in the TensorFlow repo](https://github.com/tensorflow/tensorflow/blob/3b86d3c0168338cc2c8a4947e7f02d28403c61e2/tensorflow/compiler/xla/python/pytree.h#L105) and I noticed that indeed no constructor is defined, nor any other methods that would be helpful to me.

Thanks for your help, and sorry for any confusion initially, I could have been clearer in hindsight.

**Comment by peterroelants** [comment, marked as answer]
From the discussion at: https://github.com/google/jax/discussions/13768#discussioncomment-4476533

It is not possible to create or edit `PyTreeDef` objects.

`jaxlib.xla_extension.pytree.PyTreeDef` are defined as C++ code in the [Tensorflow XLA Library](https://github.com/tensorflow/tensorflow/blob/3b86d3c0168338cc2c8a4947e7f02d28403c61e2/tensorflow/compiler/xla/python/pytree.h#L105). These do not seem to expose methods that could be useful in solving this issue.

**Comment by prasanthcakewalk** [comment]
This is probably not recommended, but FWIW there seems to be an unsupported way to create modified `PyTreeDef` objects (at least, as of writing this post), using the `PyTreeDef.from_node_data_and_children` method. Here's an MWE with some comments:

```python
import jax
import dataclasses

# Creating an example node class
@jax.tree_util.register_dataclass
@dataclasses.dataclass
class Foo:
    child_1: int
    child_2: float
    aux_1: bool = dataclasses.field(metadata={"static": True})
    aux_2: float = dataclasses.field(metadata={"static": True})

foo = Foo(1, 1.0, False, 2.0)  # create a tree
leaves, cur_treedef = jax.tree.flatten(foo)
print(leaves)  # prints [1, 1.0]
print(cur_treedef)  # prints PyTreeDef(CustomNode(Foo[(False, 2.0)], [*, *]))

_registry = jax._src.tree_util.default_registry  # This is a registry where the different pytree node types get registered to
_PyTreeDef = jax._src.tree_util.PyTreeDef  # This class contains a method that can be used for creating PyTreeDef objects

cur_node_data = cur_treedef.node_data()  # returns a tuple: first element = node-class, second element = aux-data returned by the registered unflatten method.
print(cur_node_data)  # prints (<class '__main__.Foo'>, (False, 2.0))

# Creating new node_data
new_node_data = (Foo, (True, 10.0))

# Creating a new treedef object with new_node_data
new_treedef = _PyTreeDef.from_node_data_and_children(
    registry=_registry,
    node_data=new_node_data,
    children=cur_treedef.children(),  # This is not the list of children of the original tree, but the list of treedefs of the children.
)

print(new_treedef)  # prints PyTreeDef(CustomNode(Foo[(True, 10.0)], [*, *]))

# Unflatten with the new treedef
modified_foo = jax.tree.unflatten(new_treedef, leaves)

print(modified_foo)  # prints Foo(child_1=1, child_2=1.0, aux_1=True, aux_2=10.0)
```

Some notes:
- The `default_registry` and `PyTreeDef` are not part of the public API. (It'll be cool if some public/supported API is made available in jax though.)
- `from_node_data_and_children` is a staticmethod, so `_PyTreeDef.from_node_data_and_children` can be replaced with `cur_treedef.from_node_data_and_children`.
- The flatten implementation provided via `jax.tree_util.register_dataclass` returns auxiliary information as a tuple of values. But if some custom framework returns a tuple of (name, value) pairs, editing the auxiliary information would be more transparent.
- `cur_treedef.children()` is not the list of children of the original tree, but the list of treedefs of the children.. The treedef of a tree is made of (1) the identity of the root-class, (2) auxiliary information corresponding to root-node, (3) treedefs of the children (recursively).

**Reply by prasanthcakewalk** [reply]
On second thought, `from_node_data_and_children` allows creation of brand new PyTreeDef objects from scratch, not just modified versions of existing instances.
